### PR TITLE
CDAP-8902 Add CLI endpoints for scheduling

### PR DESCRIPTION
**Pull request by Denton-L**
This commit introduces introduces 'add time schedule', 'update time
schedule' and 'delete schedule' CLI endpoints. In addition to this, a
lot of refactoring was done in order to pull out unnecessarily nested
classes.

JIRA: https://issues.cask.co/browse/CDAP-8902
Bamboo: https://builds.cask.co/browse/CDAP-RUT876

### CODE REVIEW

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/AddTimeScheduleCommand.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright © 2015-2017 Cask Data, Inc.
```
Comment (awholegunch) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/AddTimeScheduleCommand.java]: 2017, as this is a new file.

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/AddTimeScheduleCommand.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    return String.format("add time schedule <%s> for <%s> <%s> " +
+                           "[description <%s>] at <%s> [properties <%s>] [constraints <%s>]",
+                         ArgumentName.SCHEDULE, ArgumentName.PROGRAM_TYPE, ArgumentName.PROGRAM,
+                         ArgumentName.DESCRIPTION, ArgumentName.CRON_EXPRESSION, ArgumentName.SCHEDULE_PROPERTIES,
+                         ArgumentName.SCHEDULE_RUN_CONSTRAINTS);
+  }
+
+  @Override
+  public String getDescription() {
+    return String.format("Adds %s which is associated with %s given.",
```
Comment (awholegunch) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/AddTimeScheduleCommand.java]: No trailing period.:
```
with %s given",

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/UpdateTimeScheduleCommand.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    return String.format("update time schedule <%s> for <%s> <%s> " +
+                           "[description <%s>] at <%s> [properties <%s>] [constraints <%s>]",
+                         ArgumentName.SCHEDULE, ArgumentName.PROGRAM_TYPE, ArgumentName.PROGRAM,
+                         ArgumentName.DESCRIPTION, ArgumentName.CRON_EXPRESSION, ArgumentName.SCHEDULE_PROPERTIES,
+                         ArgumentName.SCHEDULE_RUN_CONSTRAINTS);
+  }
+
+  @Override
+  public String getDescription() {
+    return String.format("Updates %s which is associated with %s given.",
```
Comment (awholegunch) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/UpdateTimeScheduleCommand.java]: No trailing period.

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/AddTimeScheduleCommand.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                         Fragment.of(Article.A, ElementType.SCHEDULE.getName()),
+                         Fragment.of(Article.THE, ElementType.PROGRAM.getName()));
```
Comment (awholegunch) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/AddTimeScheduleCommand.java]: Why is "THE" in blue? Does GitHub think it is something else?

Comment (Denton-L) [reply cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/AddTimeScheduleCommand.java]: Actually, I think the `A` should be blue too but I guess their regex doesn't match single-letter all-caps variables.

The `THE` is correctly blue, though.

File: cdap-cli/src/main/java/co/cask/cdap/cli/ElementType.java [hunk 5]
```diff
@@ -57,6 +57,8 @@
 
   PROGRAM("program", new Noun("program"), new Noun("Program"), null, null, ArgumentName.PROGRAM),
 
+  SCHEDULE("schedule", new Noun("schedule"), new Noun("Schedule"), null, null, ArgumentName.SCHEDULE),
```
Comment (anew) [inline cdap-cli/src/main/java/co/cask/cdap/cli/ElementType.java:60]: please update the copyright header to 2017

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/AddTimeScheduleCommand.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    scheduleClient.add(scheduleId, configuration);
+    printStream.printf("Successfully added schedule '%s' in program '%s'\n", scheduleName, appId);
+  }
+
+  @Override
+  public String getPattern() {
+    return String.format("add time schedule <%s> for <%s> <%s> " +
+                           "[description <%s>] at <%s> [properties <%s>] [constraints <%s>]",
+                         ArgumentName.SCHEDULE, ArgumentName.PROGRAM_TYPE, ArgumentName.PROGRAM,
```
Comment (anew) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/AddTimeScheduleCommand.java]: instead of allowing different program types, we should only allow it for workflows. You can hardcode that in the command syntax for now. Please do not remove because we want to be able to support other types in the future.

File: cdap-cli/src/main/java/co/cask/cdap/cli/ArgumentName.java [hunk 7]
```diff
@@ -70,12 +71,16 @@
   QUERY("query"),
   APP("app-id"),
   APP_VERSION("app-version"),
+  PROGRAM_TYPE("program-type"),
```
Comment (anew) [inline cdap-cli/src/main/java/co/cask/cdap/cli/ArgumentName.java]: I believe this won't be needed, because we will only support workflow for now.

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/AddTimeScheduleCommand.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+  @Inject
+  public AddTimeScheduleCommand(CLIConfig cliConfig, ScheduleClient scheduleClient) {
+    super(cliConfig);
+    this.scheduleClient = scheduleClient;
+  }
+
+  @Override
+  public void perform(Arguments arguments, PrintStream printStream) throws Exception {
+    String[] scheduleIdParts = arguments.get(ArgumentName.SCHEDULE.toString()).split("\\.");
```
Comment (anew) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/AddTimeScheduleCommand.java]: the schedule id does not need to contain the app name. It is implied by the program id's app part.

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/AddTimeScheduleCommand.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    String programType = arguments.get(ArgumentName.PROGRAM_TYPE.toString()).toUpperCase();
+    String[] programIdParts = arguments.get(ArgumentName.PROGRAM.toString()).split("\\.");
+    String scheduleDescription = arguments.getOptional(ArgumentName.DESCRIPTION.toString(), "");
+    String cronExpression = arguments.get(ArgumentName.CRON_EXPRESSION.toString());
+    String schedulePropertiesString = arguments.getOptional(ArgumentName.SCHEDULE_PROPERTIES.toString(), "");
+    String scheduleRunConstraintsString = arguments.getOptional(ArgumentName.SCHEDULE_RUN_CONSTRAINTS.toString(), "");
```
Comment (anew) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/AddTimeScheduleCommand.java]: I am not a big fan of parsing this as a String map, it is too error-prone to enter a property name such as "maxConcrurentRuns" and if it has a typo, it would simply be ignored. Can we instead have a syntax such as "with concurrency N"?

Comment (anew) [reply cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/AddTimeScheduleCommand.java]: Also, can we change the order and have the runConstraints before the properties?

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/ListWorkflowSchedulesCommand.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    return schedule.getClass().getName();
+  }
+
+  private String getScheduleProperties(Schedule schedule) {
+    if (schedule instanceof TimeSchedule) {
+      TimeSchedule timeSchedule = (TimeSchedule) schedule;
+      return String.format("cron entry: %s", timeSchedule.getCronEntry());
+    } else if (schedule instanceof StreamSizeSchedule) {
+      StreamSizeSchedule streamSizeSchedule = (StreamSizeSchedule) schedule;
+      return String.format("stream: %s trigger MB: %d",
```
Comment (anew) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/ListWorkflowSchedulesCommand.java]: should there be a comma before trigger: ?

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/UpdateTimeScheduleCommand.java [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+    scheduleClient.update(scheduleId, configuration);
+    printStream.printf("Successfully updated schedule '%s' in program '%s'\n", scheduleName, appId);
+  }
+
+  @Override
+  public String getPattern() {
+    return String.format("update time schedule <%s> for <%s> <%s> " +
+                           "[description <%s>] at <%s> [properties <%s>] [constraints <%s>]",
+                         ArgumentName.SCHEDULE, ArgumentName.PROGRAM_TYPE, ArgumentName.PROGRAM,
```
Comment (anew) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/UpdateTimeScheduleCommand.java]: similar to add schedule, we should only allow program type workflow, and only require the schedule name instead of the full schedule id.

File: cdap-cli/src/main/java/co/cask/cdap/cli/commandset/ScheduleCommands.java [hunk 12]
```diff
@@ -16,245 +16,41 @@
...
+        .add(injector.getInstance(SuspendScheduleCommand.class))
+        .add(injector.getInstance(UpdateTimeScheduleCommand.class))
         .build());
   }
 
   @Override
   public String getCategory() {
     return CommandCategory.APPLICATION_LIFECYCLE.getName();
   }
-
```
Comment (anew) [inline cdap-cli/src/main/java/co/cask/cdap/cli/commandset/ScheduleCommands.java:70]: I assume that all these were simply moved out without modification and did not pay too much attention to detail. If that is not the case, please let me know.

File: cdap-cli/src/main/java/co/cask/cdap/cli/english/Article.java [hunk 13]
```diff
@@ -39,4 +40,13 @@ public String toString(Word word) {
     }
   }
 
+  /**
+   * "the"
+   */
+  private static class The extends Article {
```
Comment (anew) [inline cdap-cli/src/main/java/co/cask/cdap/cli/english/Article.java:46]: what's the point of this? It seems that it can just be the string constant "the". Maybe I am missing something.

File: cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java [hunk 14]
```diff
@@ -70,6 +74,49 @@ public ScheduleClient(ClientConfig config) {
...
+  public void add(Id.Schedule schedule, ScheduleInstanceConfiguration configuration) throws IOException,
+    UnauthenticatedException, NotFoundException, UnauthorizedException, AlreadyExistsException {
+    add(schedule.toEntityId(), configuration);
+  }
+
+  public void add(ScheduleId scheduleId, ScheduleInstanceConfiguration configuration) throws IOException,
+    UnauthenticatedException, NotFoundException, UnauthorizedException, AlreadyExistsException {
+
+    String path = String.format("apps/%s/versions/%s/schedules/%s",
+                                scheduleId.getApplication(), scheduleId.getVersion(), scheduleId.getSchedule());
```
Comment (anew) [inline cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java:79]: what if there is no version?

File: cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java [hunk 15]
```diff
@@ -70,6 +74,49 @@ public ScheduleClient(ClientConfig config) {
...
+      throw new NotFoundException(scheduleId);
+    } else if (HttpURLConnection.HTTP_CONFLICT == response.getResponseCode()) {
+      throw new AlreadyExistsException(scheduleId);
+    }
+  }
+
+  @Deprecated
+  public void update(Id.Schedule schedule, ScheduleInstanceConfiguration configuration) throws IOException,
+    UnauthenticatedException, NotFoundException, UnauthorizedException, AlreadyExistsException {
+    add(schedule, configuration);
```
Comment (anew) [inline cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java]: this should be update() not add(), right?

File: cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java [hunk 16]
```diff
@@ -70,6 +74,49 @@ public ScheduleClient(ClientConfig config) {
...
+
+  public void update(ScheduleId scheduleId, ScheduleInstanceConfiguration configuration) throws IOException,
+    UnauthenticatedException, NotFoundException, UnauthorizedException, AlreadyExistsException {
+
+    String path = String.format("apps/%s/versions/%s/schedules/%s/update",
+                                scheduleId.getApplication(), scheduleId.getVersion(), scheduleId.getSchedule());
+    URL url = config.resolveNamespacedURLV3(scheduleId.getNamespaceId(), path);
+    HttpRequest request = HttpRequest.post(url).withBody(GSON.toJson(configuration)).build();
+    HttpResponse response = restClient.execute(request, config.getAccessToken(),
+                                               HttpURLConnection.HTTP_NOT_FOUND);
```
Comment (anew) [inline cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java:100]: does it ever return bad request? For example if you update a time schedule to be a stream schedule?

Comment (anew) [reply cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java:100]: or if the cron expression is invalid?

File: cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java [hunk 17]
```diff
@@ -167,7 +214,25 @@ public void resume(ScheduleId scheduleId) throws IOException, UnauthenticatedExc
   }
 
   @Deprecated
-  public String getStatus(Id.Schedule schedule)
+  public void delete(Id.Schedule schedule)
+    throws IOException, UnauthenticatedException, NotFoundException, UnauthorizedException {
+    resume(schedule.toEntityId());
```
Comment (anew) [inline cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java]: you mean, delete()?

File: cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java [hunk 18]
```diff
@@ -70,6 +74,49 @@ public ScheduleClient(ClientConfig config) {
...
+                                               HttpURLConnection.HTTP_NOT_FOUND,
+                                               HttpURLConnection.HTTP_CONFLICT);
+    if (HttpURLConnection.HTTP_NOT_FOUND == response.getResponseCode()) {
+      throw new NotFoundException(scheduleId);
+    } else if (HttpURLConnection.HTTP_CONFLICT == response.getResponseCode()) {
+      throw new AlreadyExistsException(scheduleId);
+    }
+  }
+
+  @Deprecated
```
Comment (anew) [inline cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java]: There is no need to introduce a new deprecated method....

Comment (anew) [reply cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java]: And the same for the other methods.

File: cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java [hunk 19]
```diff
@@ -187,4 +252,26 @@ public String getStatus(ScheduleId scheduleId) throws IOException, Unauthenticat
...
+    URL url = config.resolveNamespacedURLV3(workflowId.getParent().getParent(), path);
+    HttpResponse response = restClient.execute(HttpMethod.GET, url, config.getAccessToken(),
+                                               HttpURLConnection.HTTP_NOT_FOUND);
+    if (HttpURLConnection.HTTP_NOT_FOUND == response.getResponseCode()) {
+      throw new NotFoundException(workflowId);
+    }
+
+    // TODO: Figure out how to do this the proper way using a response object
+    Gson formatterGson = new GsonBuilder().setPrettyPrinting().create();
+    return formatterGson.toJson(new String(response.getResponseBody(), Charsets.UTF_8));
```
Comment (anew) [inline cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java]: could this be formatted as a table?

**Review by anew — COMMENTED** [review]
please see my comments

File: cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java (hunk 19 above)
Comment (Denton-L) [reply cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java]: My bad; I forgot to cut out these two methods. It's being entirely removed.

File: cdap-cli/src/main/java/co/cask/cdap/cli/english/Article.java (hunk 13 above)
Comment (Denton-L) [reply cdap-cli/src/main/java/co/cask/cdap/cli/english/Article.java:46]: Some methods invoke `Fragment.of(Article.A...`. For consistency, I introduced `Article.THE` which can be used in the same way but it's not necessary. I can cut it if you think it's clearer.

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/ListWorkflowSchedulesCommand.java (hunk 10 above)
Comment (Denton-L) [reply cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/ListWorkflowSchedulesCommand.java]: I didn't write this but I can put one in.

File: cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java (hunk 16 above)
Comment (Denton-L) [reply cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java:100]: Currently, it throws a 500 error. I'll file a JIRA about this.

File: cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java (hunk 14 above)
Comment (Denton-L) [reply cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java:79]: I don't think there's a situation where this will be called without a version.

File: cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java (hunk 16 above)
Comment (Denton-L) [reply cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java:100]: https://issues.cask.co/browse/CDAP-8947

File: cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java (hunk 18 above)
Comment (Denton-L) [reply cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java]: I went ahead and removed all of the new deprecated methods.

File: cdap-cli/src/main/java/co/cask/cdap/cli/ArgumentName.java [hunk 20]
```diff
@@ -71,16 +71,16 @@
...
   VIEW("view-id"),
   HTTP_METHOD("http-method"),
   ENDPOINT("endpoint"),
   HEADERS("headers"),
   HTTP_BODY("body"),
   CRON_EXPRESSION("cron-expression"),
+  SCHEDULE_NAME("schedule-name"),
   SCHEDULE_PROPERTIES("schedule-properties"),
-  SCHEDULE_RUN_CONSTRAINTS("schedule-run-constraints"),
+  SCHEDULE_RUN_CONCURRENCY("schedule-run-concurrency"),
```
Comment (anew) [inline cdap-cli/src/main/java/co/cask/cdap/cli/ArgumentName.java]: wondering if just using "concurrency" would be better

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/UpdateTimeScheduleCommand.java [hunk 21]
```diff
@@ -93,11 +87,11 @@ public void perform(Arguments arguments, PrintStream printStream) throws Excepti
 
   @Override
   public String getPattern() {
-    return String.format("update time schedule <%s> for <%s> <%s> " +
-                           "[description <%s>] at <%s> [properties <%s>] [constraints <%s>]",
-                         ArgumentName.SCHEDULE, ArgumentName.PROGRAM_TYPE, ArgumentName.PROGRAM,
-                         ArgumentName.DESCRIPTION, ArgumentName.CRON_EXPRESSION, ArgumentName.SCHEDULE_PROPERTIES,
-                         ArgumentName.SCHEDULE_RUN_CONSTRAINTS);
+    return String.format("update time schedule <%s> for workflow <%s> " +
```
Comment (anew) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/schedule/UpdateTimeScheduleCommand.java]: shouldn't this also accept a version? As in:
```
  start workflow <app-id.workflow-id> [version <app-version>] [<runtime-args>]
    Starts a workflow. '<runtime-args>' is specified in the format 'key1=a key2=b'.
```
and the same for the add schedule.

File: cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java [hunk 22]
```diff
@@ -252,26 +227,4 @@ public String getStatus(ScheduleId scheduleId) throws IOException, Unauthenticat
...
     return responseObject.get("status");
   }
-
-  @Deprecated
-  public String getWorkflowSchedules(Id.Workflow workflow)
-    throws IOException, UnauthenticatedException, NotFoundException, UnauthorizedException {
-    return getWorkflowSchedules(workflow.toEntityId());
-  }
-
-  public String getWorkflowSchedules(WorkflowId workflowId) throws IOException, UnauthenticatedException,
```
Comment (anew) [inline cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java]: why was this removed?

File: cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java (hunk 14 above)
Comment (anew) [reply cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java:79]: With your current command syntax, that is possible. What does scheduleId.getVersion() return if you don't specify a version? Will the endpoint work with that? Or would you have to call the non-versioned endpoint?

**Review by anew — COMMENTED** [review]
Looks almost good now. Only two more comments.

File: cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java (hunk 14 above)
Comment (Denton-L) [reply cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java:79]: Currently, if you build a `ScheduleId` with no version, the version will default to `ApplicationId.DEFAULT_VERSION`. Additionally, if you call the unversioned API, internally it will use `ApplicationId.DEFAULT_VERSION` as well.

File: cdap-cli/src/main/java/co/cask/cdap/cli/ArgumentName.java [hunk 23]
```diff
@@ -106,6 +106,7 @@
   WORKFLOW_NODE("workflow-node"),
   WORKFLOW_TOKEN_SCOPE("workflow-token-scope"),
   WORKFLOW_TOKEN_KEY("workflow-token-key"),
+  WORKFLOW_VERSION("workflow-version"),
```
Comment (anew) [inline cdap-cli/src/main/java/co/cask/cdap/cli/ArgumentName.java]: You should just use the app-version that other commands use

File: cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java (hunk 14 above)
Comment (anew) [reply cdap-client/src/main/java/co/cask/cdap/client/ScheduleClient.java:79]: ok

**Review by anew — COMMENTED** [review]
One more remark. Thanks

**Comment by Denton-L** [comment]
@anew Completed the changes. If this is good, I'll squash and wait for the build to complete.

**Review by anew — APPROVED** [review]
LGTM
